src/Stratis.Bitcoin/Utilities/AverageCalculator.cs

```diff
@@ -0,0 +1,78 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    /// <summary>Calculates average value of last N added samples every time new sample is added.</summary>
+    /// <remarks>
+    /// Implementation doesn't iterate through the whole collection of samples when average value is being calculated which makes this component more optimal
+    /// in terms of performance when frequent calculation of an average value on a set of items is required.
+    /// </remarks>
+    public class AverageCalculator
+    {
+        /// <summary>Average value of supplied samples.</summary>
+        public double Average { get; private set; }
+
+        /// <summary>Samples used in calculation of the average value.</summary>
+        private CircularArray<double> samples;
+
+        /// <summary>Initializes a new instance of the <see cref="AverageCalculator"/> class.</summary>
+        /// <param name="maxSamples">Maximum amount of samples that can be used in the calculation of the average value.</param>
+        /// <exception cref="ArgumentException">Thrown if <paramref name="maxSamples"/> is less than 2.</exception>
+        public AverageCalculator(int maxSamples)
+        {
+            if (maxSamples < 2)
+                throw new ArgumentException("Minimal amount of samples is 2.");
+
+            this.Average = 0;
+            this.samples = new CircularArray<double>(maxSamples);
+        }
+
+        /// <summary>Gets the maximum amount of samples that can be used in the calculation of the average value.</summary>
+        public int GetMaxSamples()
+        {
+            return this.samples.Capacity;
+        }
+        
+        /// <summary>Sets the maximum amount of samples that can be used in the calculation of the average value.</summary>
+        /// <remarks>This is an expensive operation since it will require recreating an array of samples.</remarks>
+        public void SetMaxSamples(int maxSamples)
```

Comment:
Parameter

Reply:
Maybe even delete the whole method.

Reply:
If you want to keep the method, get rid of the Linq. But it seems weird the way how you use it from the block puller.

Reply:
Fixed